vllm/v1/core/block_pool.py

```diff
@@ -19,6 +18,95 @@
 logger = init_logger(__name__)
 
 
+class BlockLookupCache:
+    """
+    Cache of blocks that are used for prefix caching.
+    """
+
+    def __init__(self):
+        # {block_hash: KVCacheBlocks}.
+        # - Mostly block_hash maps to a single KVCacheBlock, and KVCacheBlocks
+        #   would simply be a KVCacheBlock.
+        # - Otherwise, KVCacheBlocks is a dict from {block_id: KVCacheBlock}
+        #
+        # NOTE: The union type is introduced in order to reduce GC costs from
+        # the inner dict.
+        #
+        # A cached block is a full block with a block hash that can be used
+        # for prefix caching.
+        # The cached block may be used by running requests or in the
+        # free_block_queue that could potentially be evicted.
+        # NOTE: We currently don't de-duplicate the blocks in the cache,
+        # meaning that if a block becomes full and is cached, we don't check
+        # if there is already an identical block in the cache. This is because
+        # we want to make sure the allocated block IDs won't change so that
+        # block tables are append-only.
+        self._cache: dict[BlockHashWithGroupId,
+                          Union[KVCacheBlock, dict[int, KVCacheBlock]]] = {}
+
+    def get_one_block(self,
+                      key: BlockHashWithGroupId) -> Optional[KVCacheBlock]:
+        """
+        Gets any block with the given block hash key.
+        """
+        blocks = self._cache.get(key)
+        if blocks is None:
+            return None
+        if type(blocks) is KVCacheBlock:
+            return blocks
+        if type(blocks) is dict:
+            return next(iter(blocks.values()))
+        self._unexpected_blocks_type(blocks)
+        return None
+
+    def insert(self, key: BlockHashWithGroupId, block: KVCacheBlock) -> None:
+        """
+        Inserts the KVCacheBlock to the cache
+        """
+        blocks = self._cache.get(key)
+        if blocks is None:
+            # When key is not found, attach a single block to the key
+            self._cache[key] = block
+        elif type(blocks) is KVCacheBlock:
+            # If there's a block with the same key, merge the original block
+            # and the new block into a dict
+            self._cache[key] = {blocks.block_id: blocks, block.block_id: block}
+        elif type(blocks) is dict:
+            # If it's already a dict, simply insert the block
+            blocks[block.block_id] = block
+        else:
+            self._unexpected_blocks_type(blocks)
+
+    def check_block_hash_and_pop_block(self, key: BlockHashWithGroupId,
+                                       block_id: int) -> bool:
```

Comment:
What about 
```
 def pop(self, key: BlockHashWithGroupId,
                                       block_id: int) -> Optional[KVCacheBlock]:
```
This is more similar to the standard dict API.

Reply:
Alternatively if we only need to return `bool`, could call it `remove`.